File: server/loans/core/src/main/java/ee/rsx/kata/bank/loans/usecases/LoanEligibilityCalculation.java
package ee.rsx.kata.bank.loans.usecases;

import ee.rsx.kata.bank.loans.domain.loan.LoanEligibility;
import ee.rsx.kata.bank.loans.domain.limits.gateway.DetermineEligiblePeriod;
import ee.rsx.kata.bank.loans.domain.loan.Applicant;
import ee.rsx.kata.bank.loans.domain.loan.Loan;
import ee.rsx.kata.bank.loans.domain.segment.CreditSegment;
import ee.rsx.kata.bank.loans.domain.segment.gateway.FindCreditSegment;
import ee.rsx.kata.bank.loans.domain.ssn.SocialSecurityNumber;
import ee.rsx.kata.bank.loans.eligibility.CalculateLoanEligibility;
import ee.rsx.kata.bank.loans.eligibility.LoanEligibilityRequestDTO;
import ee.rsx.kata.bank.loans.eligibility.LoanEligibilityResultDTO;
import ee.rsx.kata.bank.loans.validation.limits.LoadValidationLimits;
import ee.rsx.kata.bank.loans.validation.limits.ValidationLimitsDTO;
import ee.rsx.kata.bank.loans.validation.ssn.ValidateSocialSecurityNumber;
import jakarta.annotation.Nullable;
import jakarta.inject.Named;
import lombok.RequiredArgsConstructor;

import java.util.List;
import java.util.Optional;
import java.util.stream.Stream;

import static ee.rsx.kata.bank.loans.eligibility.LoanEligibilityStatus.*;
import static ee.rsx.kata.bank.loans.validation.ssn.ValidationStatus.OK;
import static java.util.Optional.*;
import static org.apache.commons.collections4.CollectionUtils.isEmpty;

@Named
@RequiredArgsConstructor
class LoanEligibilityCalculation implements CalculateLoanEligibility {

  private final ValidateSocialSecurityNumber validateSocialSecurityNumber;
  private final LoadValidationLimits loadValidationLimits;
  private final FindCreditSegment findCreditSegment;
  private final DetermineEligiblePeriod determineEligiblePeriod;

  @Override
  public LoanEligibilityResultDTO on(LoanEligibilityRequestDTO eligibilityRequest) {
    var limits = loadValidationLimits.invoke();
    List<String> validationErrors = validate(eligibilityRequest, limits);

    var eligibility = isEmpty(validationErrors)
      ? calculateEligibility(eligibilityRequest, limits)
      : new LoanEligibility(INVALID, null, null);

    return new LoanEligibilityResultDTO(
      eligibility.status(),
      validationErrors,
      eligibilityRequest.ssn(),
      eligibilityRequest.loanAmount(),
      eligibilityRequest.loanPeriodMonths(),
      eligibility.eligibleAmount(),
      eligibility.eligiblePeriod()
    );
  }

  private @Nullable List<String> validate(
    LoanEligibilityRequestDTO eligibilityRequest,
    ValidationLimitsDTO limits
  ) {
    List<String> errors = Stream.of(
        checkForSsnErrorIn(eligibilityRequest),
        checkForAmountErrorIn(eligibilityRequest, limits),
        checkForPeriodErrorIn(eligibilityRequest, limits)
      )
      .flatMap(Optional::stream)
      .toList();


    return errors.isEmpty() ? null : errors;
  }

  private Optional<String> checkForSsnErrorIn(LoanEligibilityRequestDTO request) {
    var ssnValidity = validateSocialSecurityNumber.on(request.ssn()).status();

    return ssnValidity == OK ? empty() : of("SSN is not valid");
  }

  private Optional<String> checkForAmountErrorIn(LoanEligibilityRequestDTO request, ValidationLimitsDTO limits) {
    return switch (request.loanAmount()) {
      case Integer amount when amount < limits.minimumLoanAmount() -> of("Loan amount is less than minimum required");
      case Integer amount when amount > limits.maximumLoanAmount() -> of("Loan amount is more than maximum allowed");
      default -> empty();
    };
  }

  private Optional<String> checkForPeriodErrorIn(LoanEligibilityRequestDTO request, ValidationLimitsDTO limits) {
    return switch (request.loanPeriodMonths()) {
      case Integer period when period < limits.minimumLoanPeriodMonths() -> of("Loan period is less than minimum required");
      case Integer period when period > limits.maximumLoanPeriodMonths() -> of("Loan period is more than maximum allowed");
      default -> empty();
    };
  }

  private LoanEligibility calculateEligibility(LoanEligibilityRequestDTO request, ValidationLimitsDTO limits) {
    var ssn = new SocialSecurityNumber(request.ssn());
    var amount = request.loanAmount();
    var period = request.loanPeriodMonths();

    return findCreditSegment.forPerson(ssn)
      .map(segment -> applicantWith(ssn, segment, amount, limits, period))
      .filter(Applicant::isNotInDebt)
      .map(applicant ->
          applicant.attemptFindingEligibilityWithAmount().orElseGet(() -> recalculateEligibilityFor(applicant, limits)
        ))
      .orElseGet(
        () -> new LoanEligibility(DENIED, null, null)
      );
  }

  private static Applicant applicantWith(
    SocialSecurityNumber ssn, CreditSegment segment, Integer amount, ValidationLimitsDTO limits, Integer period
  ) {
    var loan = Loan.with(amount, period, limits.minimumLoanAmount(), limits.maximumLoanAmount());
    return Applicant.with(ssn, segment, loan);
  }

  private LoanEligibility recalculateEligibilityFor(Applicant applicant, ValidationLimitsDTO limits) {
    Optional<Integer> newPeriod =
      determineEligiblePeriod.forLoan(applicant.loan().amount(), applicant.segment())
        .filter(period ->
          limits.minimumLoanPeriodMonths() <= period && period <= limits.maximumLoanPeriodMonths()
        );

    return newPeriod.map(applicant::eligibilityForPeriod)
      .orElseGet(applicant::eligibility);
  }
}

File: server/loans/core/src/main/kotlin/ee/rsx/kata/bank/loans/usecases/LoanEligibilityCalculation.kt
package ee.rsx.kata.bank.loans.usecases

import ee.rsx.kata.bank.loans.domain.limits.gateway.DetermineEligiblePeriod
import ee.rsx.kata.bank.loans.domain.limits.gateway.LoanEligibility
import ee.rsx.kata.bank.loans.domain.segment.CreditSegment
import ee.rsx.kata.bank.loans.domain.segment.gateway.FindCreditSegment
import ee.rsx.kata.bank.loans.domain.ssn.SocialSecurityNumber
import ee.rsx.kata.bank.loans.eligibility.CalculateLoanEligibility
import ee.rsx.kata.bank.loans.eligibility.LoanEligibilityRequestDTO
import ee.rsx.kata.bank.loans.eligibility.LoanEligibilityResultDTO
import ee.rsx.kata.bank.loans.eligibility.LoanEligibilityStatus
import ee.rsx.kata.bank.loans.eligibility.LoanEligibilityStatus.APPROVED
import ee.rsx.kata.bank.loans.eligibility.LoanEligibilityStatus.DENIED
import ee.rsx.kata.bank.loans.eligibility.LoanEligibilityStatus.INVALID
import ee.rsx.kata.bank.loans.extensions.ifTrue
import ee.rsx.kata.bank.loans.validation.limits.LoadValidationLimits
import ee.rsx.kata.bank.loans.validation.limits.ValidationLimitsDTO
import ee.rsx.kata.bank.loans.validation.ssn.ValidateSocialSecurityNumber
import ee.rsx.kata.bank.loans.validation.ssn.ValidationStatus.OK
import jakarta.annotation.Nullable
import jakarta.inject.Named
import kotlin.math.min

@Named
internal class LoanEligibilityCalculation(
  private val validateSocialSecurityNumber: ValidateSocialSecurityNumber,
  private val loadValidationLimits: LoadValidationLimits,
  private val findCreditSegment: FindCreditSegment,
  private val determineEligiblePeriod: DetermineEligiblePeriod
) : CalculateLoanEligibility {

  override fun invoke(eligibilityRequest: LoanEligibilityRequestDTO): LoanEligibilityResultDTO {
    val limits = loadValidationLimits()
    val validationErrors = validate(eligibilityRequest, limits)

    val (status, eligibleAmount, eligiblePeriod) =
      if (validationErrors.isNullOrEmpty()) calculateEligibility(
        eligibilityRequest,
        limits
      ) else
        LoanEligibility(INVALID, null, null)

    return LoanEligibilityResultDTO(
      status,
      validationErrors,
      eligibilityRequest.ssn,
      eligibilityRequest.loanAmount,
      eligibilityRequest.loanPeriodMonths,
      eligibleAmount,
      eligiblePeriod
    )
  }

  @Nullable
  private fun validate(
    eligibilityRequest: LoanEligibilityRequestDTO,
    limits: ValidationLimitsDTO
  ): List<String>? =
    listOfNotNull(
      checkForSsnErrorIn(eligibilityRequest),
      checkForAmountErrorIn(eligibilityRequest, limits),
      checkForPeriodErrorIn(eligibilityRequest, limits)
    )
      .ifEmpty { null }

  private fun checkForSsnErrorIn(request: LoanEligibilityRequestDTO) =
    (validateSocialSecurityNumber(request.ssn).status != OK)
      .ifTrue { "SSN is not valid" }

  private fun checkForAmountErrorIn(request: LoanEligibilityRequestDTO, limits: ValidationLimitsDTO) =
    request.loanAmount.let {
      (it < limits.minimumLoanAmount).ifTrue { "Loan amount is less than minimum required" }
        ?: (it > limits.maximumLoanAmount).ifTrue { "Loan amount is more than maximum allowed" }
    }


  private fun checkForPeriodErrorIn(request: LoanEligibilityRequestDTO, limits: ValidationLimitsDTO) =
    request.loanPeriodMonths.let {
      (it < limits.minimumLoanPeriodMonths).ifTrue { "Loan period is less than minimum required" }
        ?: (it > limits.maximumLoanPeriodMonths).ifTrue { "Loan period is more than maximum allowed" }
    }

  private fun calculateEligibility(request: LoanEligibilityRequestDTO, limits: ValidationLimitsDTO): LoanEligibility {
    val ssn = SocialSecurityNumber(request.ssn)
    val creditSegment: CreditSegment? = findCreditSegment(forPerson = ssn)

    return creditSegment
      ?.let {
        val status = determineEligibilityStatusFor(request, it)

        determineEligibleAmountFor(limits, it, request.loanPeriodMonths)
          ?.let { eligibleAmount ->
            LoanEligibility(status, eligibleAmount)
          }
          ?: run { recalculateEligibilityFor(request, limits, creditSegment, status) }
      }
      ?: LoanEligibility(status = DENIED)
  }

  private fun recalculateEligibilityFor(
    request: LoanEligibilityRequestDTO,
    limits: ValidationLimitsDTO,
    segment: CreditSegment,
    status: LoanEligibilityStatus
  ): LoanEligibility {

    if (segment.isDebt) return LoanEligibility(status)

    val newPeriod = determineEligiblePeriod(forAmount = request.loanAmount, forSegment = segment)
      ?.takeIf {
        it in limits.minimumLoanPeriodMonths..limits.maximumLoanPeriodMonths
      }
    val newAmount = newPeriod?.let { determineEligibleAmountFor(limits, segment, it) }

    return LoanEligibility(status, newAmount, newPeriod)
  }

  private fun determineEligibilityStatusFor(
    request: LoanEligibilityRequestDTO, creditSegment: CreditSegment
  ): LoanEligibilityStatus {
    val creditScore = creditSegment.creditModifier.toDouble() / request.loanAmount * request.loanPeriodMonths

    return if (!creditSegment.isDebt && creditScore > 1)
      APPROVED
    else
      DENIED
  }

  private fun determineEligibleAmountFor(
    limits: ValidationLimitsDTO, creditSegment: CreditSegment, loanPeriodMonths: Int
  ): Int? {
    val eligibleAmount = min(
      limits.maximumLoanAmount.toDouble(),
      (creditSegment.creditModifier * loanPeriodMonths - 1).toDouble()
    ).toInt()

    return (eligibleAmount >= limits.minimumLoanAmount).ifTrue { eligibleAmount }
  }
}
